### PR TITLE
plist: create var paths

### PR DESCRIPTION
**Pull request by DomT4**
Ensure the ` var ` paths are created in-formulae, rather than allowing ` launchctl ` to create the paths once loaded. This should prevent a few of the occasions where the ` var ` path is created with ` root ` ownership rather than ` $USER `.

### CODE REVIEW

**Comment by DomT4** [comment]
Re: @mikemcquaid

File: Library/Formula/beanstalk.rb [hunk 1]
```diff
@@ -12,6 +10,7 @@ class Beanstalk < Formula
   end
 
   def install
+    mkdir_p var/"log"
```
Comment (mikemcquaid) [inline Library/Formula/beanstalk.rb]: Maybe these should all be in post_install?

Comment (DomT4) [inline Library/Formula/beanstalk.rb]: I wasn't sure either way. It seems to be regularly placed in ` install `, and only placed in `postinstall` when `postinstall` exists for some other reason. Wasn't sure if that was a deliberate maintainer choice over time, or.

Comment (mikemcquaid) [inline Library/Formula/beanstalk.rb]: Try and see if they are included in bottles if that are in `install`. Otherwise move them all to `post_install`

Comment (DomT4) [inline Library/Formula/beanstalk.rb]: :+1: - Will yank a bottle off the bot and take a peek, or bottle something locally and check. Whichever proves quicker.

**Comment by xu-cheng** [comment]
Can we implement such feature in the core rather than per-formula basis? Maybe change [`install_plist` method](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/formula_installer.rb#L559-L567) to this:
```ruby
  def install_plist
    return unless formula.plist
    formula.plist_path.atomic_write(formula.plist)
    formula.plist_path.chmod 0644
    log = var/"log"
    log.mkpath if formula.plist.include? log.to_s
  rescue Exception => e
    onoe "Failed to install plist file"
    ohai e, e.backtrace if debug?
    Homebrew.failed = true
  end
```

**Comment by DomT4** [comment]
> Can we implement such feature in the core rather than per-formula basis?

Discussed an audit check, which is a similar global method, and Mike wasn't terribly keen given the relatively few formulae this problem affects (Less than 10% of all ` plist ` using formulae), but happy to defer to him if he prefers such a method.

**Comment by mikemcquaid** [comment]
@xu-cheng That could be OK.

**Comment by DomT4** [comment]
Heh. Jenkins died. If we want to go the global method I'll drop this PR in the trash. Let me know either way.

**Comment by DomT4** [comment]
If we go global though, ` collectd ` will need fixing - Otherwise we'll be creating a directory in ` /usr/local/var ` even for Homebrew uses who install their Homebrew elsewhere. The formula's plist has been written ` /usr/local`-specific for some reason.

**Comment by mikemcquaid** [comment]
> If we go global though, collectd will need fixing - Otherwise we'll be creating a directory in /usr/local/var even for Homebrew uses who install their Homebrew elsewhere. The formula's plist has been written /usr/local-specific for some reason.

:+1: 

@xu-cheng Are you happy to do the core-fixup?

**Comment by xu-cheng** [comment]
:+1: for core fixup. We can leave `collectd` as it is. If I understand correctly, we shouldn't care about the permission of `/usr/local/var` unless it's in homebrew prefix.

**Comment by DomT4** [comment]
> We can leave collectd as it is. If I understand correctly, we shouldn't care about the permission of /usr/local/var unless it's in homebrew prefix.

Yes, but we shouldn't be *creating* directories in ` /usr/local ` if that's not the ` HOMEBREW_PREFIX `. [This blob](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/collectd.rb#L70-L73) does.

**Comment by xu-cheng** [comment]
> but we shouldn't be creating directories in /usr/local if that's not the HOMEBREW_PREFIX. This blob does.

My implementation only create `formula.var/"log"` if it's mentioned in the plist file. See https://github.com/Homebrew/homebrew/pull/37967

**Comment by DomT4** [comment]
Yes, but ` launchctl ` *won't* obey Homebrew's sandbox. If the ` plist ` identifies the log path as `/Users/Dogs/DogsAreFurry ` launchctl will create that path as soon as the ` plist ` is loaded.

**Comment by xu-cheng** [comment]
Sorry, we are not in the same page. I thought `collectd` have to work in `/usr/local/var` for some reason. Feel free to create a PR to fix it. :wink:

**Pull request closed without merging**